### DISCUSSION TITLE
Turning inlined varargs into a generic tuple

### DISCUSSION THREAD

**Discussion by soronpo** [scala/scala3#12651] [Metaprogramming] [answered]
I've tried to do this, but I'm hitting a wall. How can I convert varargs to a tuple generically, without loosing the type narrowness.
E.g.:
```scala
transparent inline def toTuple(inline args : Int*) : Tuple 

val x : (1, 2, 3) = toTuple(1,2,3)
```

**Comment by nicolasstucki** [comment, marked as answer]
```scala
import scala.quoted.*

def toTupleExpr(argsExpr: Expr[Seq[Int]])(using Quotes): Expr[Tuple] = {
  argsExpr match
    case Varargs(argExprs) =>
      argExprs.foldRight[Expr[Tuple]]('{EmptyTuple}) { (expr, acc) =>
        (expr, acc) match // match to extract precise type
          case ('{ $expr: head }, '{ type tail <: Tuple; $acc: `tail` }) =>
            '{ $expr *: $acc }
      }
    case _ =>
      quotes.reflect.report.throwError(
        "Expected explicit argument list but got: " + argsExpr.show, argsExpr)
}
```

```scala
transparent inline def toTuple(inline args : Int*) : Tuple =
 ${ toTupleExpr('args) }

val x : (1, 2, 3) = toTuple(1,2,3)
```

**Reply by nicolasstucki** [reply]
I noticed that some of the tuple optimizations to convert `1 *: 2 *: 3 *: EmptyTuple` into `Tuple3(1, 2, 3)` are not kicking in.

**Reply by soronpo** [reply]
> I noticed that some of the tuple optimizations to convert `1 *: 2 *: 3 *: EmptyTuple` into `Tuple3(1, 2, 3)` are not kicking in.

Scala is an eager widener :wink: 

Do you think this method has a place in the official compiler library as a varargs apply method under the `Tuple` object?

**Reply by nicolasstucki** [reply]
I would first create a library with it and see how many use cases there are.
